### ISSUE TITLE
Divine Rapier tracker

### ISSUE THREAD

**Issue by evaryont** [yasp-dota/yasp#333]
Much like #322, it would be a neat thing for games that actually do get one:

- Who bought the rapier(s)?
- Did they ever drop?
- Who picked them up?

Especially useful for clowny games that result in multiple rapiers switching between teams.

**Comment by howardc93** [comment]
You can see who bought rapiers in the item purchases timeline.  I don't know if there's a better way of telling when they dropped than just "next death" of the player.  I also don't know how to figure out who picked them up.

**Comment by McHearty** [comment]
You should be able to track Rapier, as well as sprees and aegis through DOTA_TournamentEvents if I'm not mistaken, don't believe this works for gem though.

**Comment by howardc93** [comment]
I guess this really hinges on figuring out how to get inventory items in the replay.  

Maybe they're stored within the entity of the hero the player controls?

@spheenik might know

**Comment by spheenik** [comment]
I have some code that reads what is in player inventories.
Gonna dig for it this evening.

**Comment by spheenik** [comment]
Better late than never, does this help?

https://gist.github.com/spheenik/c63eaeaadbeb01428144

**Comment by howardc93** [comment]
Yeah, it looks like it'll need to be adapted for clarity 2, and I think I can drop the bottle and rune cases, but it's a good starting point

**Comment by howardc93** [comment]
We would be able to track time holding gem/rapier for each player.  There's no way to "trace" the ownership path of a particular one though

**Comment by howardc93** [comment]
also, it would have been funny if this was issue #322

**Comment by spheenik** [comment]
If you track the rapiers handle, you can know who picked it up.

**Comment by howardc93** [comment]
Ah, I see, each inventory item is an distinct entity.

So every second, we check each player's inventory (inventory slots 0-5?) and emit the names of current items (as an array?) along with other player state such as lh/gold/xp/position.

Then we can count the number of seconds each player held an item such as rapier/gem (or any other item, like wards).

If we want to trace a particular rapier/gem. . . When we encounter an item with name gem/rapier, we place that handle in a hash table with key=handle, value=parent entity handle.

When we do our every-second check of items, we see if a handle appears in the "tracked" items list, and we can emit an event if the "owner" entity has changed?

Do item entities have a "parent" we can use so we can detect drops?

**Comment by spheenik** [comment]
This should work!
I don't know if an item has a "current holder" handle. You might want to check with the dtInspector (clarity-examples) to list all available properties.

**Comment by howardc93** [comment]
I'm thinking of holding off on this until Source 2.  The replay format may change, and we have other features to work on.

**Comment by spheenik** [comment]
It will definitely change!

**Comment by howardchung** [comment]
with https://github.com/odota/parser/pull/6

The parser log entries now contain hero_inventory so we can count the number seconds each player held gem/rapier.
